doc/developer/design/19042023_stash_migrations.md

```diff
@@ -0,0 +1,352 @@
+
+Along with maintaining `const STASH_VERSION` and using a `match` statement as described above we'd
+setup a pattern of defining migrations in their own modules. Specifically we'd introduce the 
+following layout in our repo:
+
+```
+/stash
+    /src
+        /migrations
+            v0_to_v1.rs
+            v1_to_v2.rs
+            v2_to_v3.rs
+            ...
+        ...
+    /proto
+        ...
+```
+
+and our migration code would then look something like this:
+
+```
+const STASH_VERSION: u64 = 3;
+
+async fn migrate(...) -> Result<...> {
+    let version = stash.read_version().await?;
+
+    const NEXT_VERSION = STASH_VERSION + 1;
+    match version {
+        0 => migrations::v0_to_v1::migrate(...).await?;
+        1 => migrations::v1_to_v2::migrate(...).await?;
+        2 => migrations::v2_to_v3::migrate(...).await?;
+        3 => // migrate
+        NEXT_VERSION.. => panic!("Traveled to the future!"),
+    }
+}
+```
+
+There are a few benefits to putting each migration function in their own modules:
+
+1. The only place we would import the generated protobuf code from our snapshots would be in these
+modules. So the only place that would import the `object_v1.proto` would be the `v0_to_v1` and 
+`v1_to_v2` modules. This helps prevent us from depending on the snapshotted versions anywhere
+besides the necessary migration paths.
+2. We promote a pattern of documenting the migration and why it exists in a module doc comment.
+While it's possible to have doc comments on functions, I believe it would be more obvious/easier to
+review module level doc comments.
+
+Each migration function would be provided a `stash::Transaction`, which would allow us to open 
+[`StashCollection`](https://github.com/MaterializeInc/materialize/blob/23d7cad1211b4d1d1de3a26652937f9f2c4df61c/src/stash/src/transaction.rs#L136)s
+which should provide enough flexibility to facilitate any migration.
+
+> Note: `StashCollection`s do not enforce uniqueness like [`TableTransaction`](https://github.com/MaterializeInc/materialize/blob/23d7cad1211b4d1d1de3a26652937f9f2c4df61c/src/stash/src/lib.rs#L775-L793)s
+do, some more thought is needed here as to whether or not uniqueness checks are required for
+migrations. They probably are?
+
+
+# Rollout
+[rollout]: #rollout
+
+At our current scale, I don't believe the benefit to partially rolling out a new format for the
+Stash outweighs the complexity of concurrently maintaining two separate implementations. To 
+get similar testing coverage that a partial rollout would, we can validate that the new Stash
+format and migration work, by running the `stash-debug`'s `upgrade-check` command against select
+customer environments.
+
+
+## Testing and observability
+[testing-and-observability]: #testing-and-observability
+
+This change will largely be tested manually, and via Rust tests. I don't believe there are any
+sqllogictest tests or testdrive tests that would be able to exercise this change in a unique way.
```

Comment:
What about the existing upgrade tests that we have? Would those work?

https://github.com/MaterializeInc/materialize/tree/main/test/upgrade

Reply:
Yes, those would actually be great! I updated the doc to include these